src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -238,12 +238,65 @@ enum AnchorFailure {
     RustdocAnchorConflict(Res),
 }
 
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+crate enum UrlFragment {
+    Method(Symbol),
+    TyMethod(Symbol),
+    AssociatedConstant(Symbol),
+    AssociatedType(Symbol),
+
+    StructField(Symbol),
+    Variant(Symbol),
+    VariantField { variant: Symbol, field: Symbol },
+
+    UserWritten(String),
+}
+
+impl UrlFragment {
+    /// Create a fragment for an associated item.
+    ///
+    /// `is_prototype` is whether this associated item is a trait method
+    /// without a default definition.
+    fn from_assoc_item(name: Symbol, kind: ty::AssocKind, is_prototype: bool) -> Self {
+        match kind {
+            ty::AssocKind::Fn => {
+                if is_prototype {
+                    UrlFragment::TyMethod(name)
+                } else {
+                    UrlFragment::Method(name)
+                }
+            }
+            ty::AssocKind::Const => UrlFragment::AssociatedConstant(name),
+            ty::AssocKind::Type => UrlFragment::AssociatedType(name),
+        }
+    }
+}
+
+/// Render the fragment, including the leading `#`.
+impl std::fmt::Display for UrlFragment {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```

Comment:
I thought adding a `Display` impl would be the best way to go, but I could be convinced of changing this to a method that takes `&mut String` or similar because

- it would then have an explanatory method name (e.g., make it clear that it includes the `#`)
- it might be more efficient since it avoids the `Result` and using `Formatter`